mainargs/src-3/Macros.scala

```diff
@@ -0,0 +1,152 @@
+package mainargs
+
+import scala.quoted._
+
+object Macros {
+  private def mainAnnotation(using Quotes) = quotes.reflect.Symbol.requiredClass("mainargs.main")
+  private def argAnnotation(using Quotes) = quotes.reflect.Symbol.requiredClass("mainargs.arg")
+  def parserForMethods[B](base: Expr[B])(using Quotes, Type[B]): Expr[ParserForMethods[B]] = {
+    import quotes.reflect._
+    val allMethods = TypeRepr.of[B].typeSymbol.memberMethods
+    val annotatedMethodsWithMainAnnotations = allMethods.flatMap { methodSymbol =>
+      methodSymbol.getAnnotation(mainAnnotation).map(methodSymbol -> _)
+    }.sortBy(_._1.pos.map(_.start))
+    val mainDatas = Expr.ofList(annotatedMethodsWithMainAnnotations.map { (annotatedMethod, mainAnnotationInstance) =>
+      createMainData[Any, B](annotatedMethod, mainAnnotationInstance)
+    })
+
+    '{
+      new ParserForMethods[B](
+        MethodMains[B]($mainDatas, () => $base)
+      )
+    }
+  }
+
+  def parserForClass[B](using Quotes, Type[B]): Expr[ParserForClass[B]] = {
+    import quotes.reflect._
+    val typeReprOfB = TypeRepr.of[B]
+    val companionModule = typeReprOfB match {
+      case TypeRef(a,b) => TermRef(a,b)
+    }
+    val typeSymbolOfB = typeReprOfB.typeSymbol
+    val companionModuleType = typeSymbolOfB.companionModule.tree.asInstanceOf[ValDef].tpt.tpe.asType
+    val companionModuleExpr = Ident(companionModule).asExpr
+    val mainAnnotationInstance = typeSymbolOfB.getAnnotation(mainAnnotation).getOrElse {
+      report.throwError(
```

Comment:
@lolgab why did you make it throw here? the main annotation is not necessary if you look at the Scala 2 implementation - and a default is provided

Reply:
It was probably an overlook on my side. If you can change it to match the Scala 2 behavior we can release a new version, so you can continue with Mill's Scala 3 upgrade.

Reply:
alright, no worries :)